src/target/riscv/riscv-013.c

```diff
@@ -2630,15 +2654,58 @@ static int sb_write_address(struct target *target, target_addr_t address,
 static int batch_run(struct target *target, struct riscv_batch *batch)
 {
 	RISCV_INFO(r);
-	const int result = riscv_batch_run(batch, /*resets_delays*/  r->reset_delays_wait >= 0,
+	riscv_batch_add_nop(batch);
+	const int result = riscv_batch_run_from(batch, 0,
+			/*resets_delays*/  r->reset_delays_wait >= 0,
 			r->reset_delays_wait);
-	/* TODO: `finished_scans` should be the number of scans that have
-	 * finished, not the number of scans scheduled. */
+	/* TODO: To use `riscv_batch_finished_scans()` here, it is needed for
+	 * all scans to not discard input, meaning
+	 * "riscv_batch_add_dm_write(..., false)" should not be used. */
 	const size_t finished_scans = batch->used_scans;
 	decrement_reset_delays_counter(target, finished_scans);
 	return result;
 }
 
+/* It is expected that during creation of the batch
+ * "riscv_batch_add_dm_write(..., false)" was not used. */
```

Comment:
There is an issue with `riscv_batch_add_dm_write(..., /*read_back*/ false)` -- it discards the value read from the target.
Therefore the approach of "the next scan returns the result of the previous one" does not work.
Using `riscv_batch_add_dm_write(..., false)` requires another mechanism to determine which scan resulted in a busy response.

One approach would be to eliminate the `read_back` argument of `riscv_batch_add_dm_write()` and always read the response.
However, this is not desirable. Discarding the response significantly reduces USB traffic, which quite often is the bottleneck limiting the performance. And `riscv_batch_add_dm_write(..., /*read_back*/ false)` is exactly what is used in the most time-consuming operations -- memory writes.
E.g. when writing via Program Buffer there is no need to rely on DMI busy check for each scan. The DMI busy status can be determined by the response to the last NOP in the batch. And once it is known that the batch encountered DMI busy, OpenOCD reads target's register that contains the address of the element on which the busy status was encountered. An added benefit is that the algorithm is the same for `abstractcs.cmderr == busy`.

However, IMHO, adapting this functions to be usable in such cases can be beneficial. This can be done by the following changes:
`riscv_batch_was_scan_busy()` should use a callback in case `field->in_value` is zero. This callback will be optionally specified on batch allocation. The callback is specific to the content of the batch.
@JanMatCodasip, please let me know your thoughts on this.

Reply:
Thank you for explanation.

> One approach would be to eliminate the read_back argument of riscv_batch_add_dm_write() and always read the response. However, this is not desirable.

I fully agree. The `read_back` argument was introduced for a good reason (reduction of traffic to the JTAG adapter) and I believe it should be kept.

> (...) This can be done by the following changes: riscv_batch_was_scan_busy() should use a callback in case field->in_value is zero. This callback will be optionally specified on batch allocation. The callback is specific to the content of the batch.

For simplicity of the batch API, I would recommend to not introduce any callbacks. Instead, I would strongly prefer to leave the responsibility on the caller to determine what part of the batch needs to be re-executed. That is, the caller inspects the output from the batch and decide if and what needs to re-execute. Then it simply calls `riscv_batch_run_from()` with the appropriate `start_idx` value.